proof.go

```diff
@@ -465,6 +465,202 @@ func (proof Proof) VerifyInclusion(h hash.Hash, nid namespace.ID, leavesWithoutN
+// The subtreeWidth is also defined in ADR-013.
+// More information on the algorithm used can be found in the ToLeafRanges() method docs.
+func (proof Proof) VerifySubtreeRootInclusion(nth *NmtHasher, subtreeRoots [][]byte, subtreeWidth int, root []byte) (bool, error) {
+	// check that the proof range is valid
+	if proof.Start() < 0 || proof.Start() >= proof.End() {
+		return false, fmt.Errorf("proof range [proof.start=%d, proof.end=%d) is not valid: %w", proof.Start(), proof.End(), ErrInvalidRange)
+	}
+
+	// check that the root is valid w.r.t the NMT hasher
+	if err := nth.ValidateNodeFormat(root); err != nil {
+		return false, fmt.Errorf("root does not match the NMT hasher's hash format: %w", err)
+	}
+	// check that all the proof.Notes() are valid w.r.t the NMT hasher
+	for _, node := range proof.Nodes() {
+		if err := nth.ValidateNodeFormat(node); err != nil {
+			return false, fmt.Errorf("proof nodes do not match the NMT hasher's hash format: %w", err)
+		}
+	}
+	// check that all the subtree roots are valid w.r.t the NMT hasher
+	for _, subtreeRoot := range subtreeRoots {
+		if err := nth.ValidateNodeFormat(subtreeRoot); err != nil {
+			return false, fmt.Errorf("inner nodes does not match the NMT hasher's hash format: %w", err)
+		}
+	}
+
+	// get the subtree roots leaf ranges
+	ranges, err := ToLeafRanges(proof.Start(), proof.End(), subtreeWidth)
+	if err != nil {
+		return false, err
+	}
+
+	// check whether the number of ranges matches the number of subtree roots.
+	// if not, make an early return.
+	if len(subtreeRoots) != len(ranges) {
+		return false, fmt.Errorf("number of subtree roots %d is different than the number of the expected leaf ranges %d", len(subtreeRoots), len(ranges))
+	}
+
+	var computeRoot func(start, end int) ([]byte, error)
+	// computeRoot can return error iff the HashNode function fails while calculating the root
+	computeRoot = func(start, end int) ([]byte, error) {
+		// reached a leaf
+		if end-start == 1 {
+			// if the leaf index falls within the proof range, pop and return a
+			// leaf
+			if proof.Start() <= start && start < proof.End() {
+				// this check should always be the case.
+				// however, it is added to avoid nil pointer exceptions
+				if len(ranges) != 0 {
+					// advance the list of ranges
+					ranges = ranges[1:]
+				}
+				// advance leafHashes
+				return popIfNonEmpty(&subtreeRoots), nil
+			}
+
+			// if the leaf index is outside the proof range, pop and return a
+			// proof node (which in this case is a leaf) if present, else return
+			// nil because leaf doesn't exist
+			return popIfNonEmpty(&proof.nodes), nil
+		}
+
+		// if the current range does not overlap with the proof range, pop and
+		// return a proof node if present, else return nil because subtree
+		// doesn't exist
+		if end <= proof.Start() || start >= proof.End() {
+			return popIfNonEmpty(&proof.nodes), nil
+		}
+
+		if len(ranges) != 0 && ranges[0].Start == start && ranges[0].End == end {
```

Comment:
Regardless of refactoring the code or not, WDYT about applying this [comment](https://github.com/celestiaorg/nmt/pull/260#discussion_r1666102418) in this part?

Reply:
[70bc2b0](https://github.com/celestiaorg/nmt/pull/260/commits/70bc2b03158aeb4d82c6de5b30810fae52316a09)